File: src/algorithms/trees/bst-operations/bst-validation/sources/bst-validation.ts
// BST Validation (Recursive) — validate BST property using min/max bounds

interface BSTNode {
  value: number;
  left: BSTNode | null;
  right: BSTNode | null;
}

function bstValidation(root: BSTNode | null): boolean {
  function validate(node: BSTNode | null, minVal: number, maxVal: number): boolean {
    if (node === null) return true; // @step:initialize

    if (node.value <= minVal || node.value >= maxVal) {
      // Node value violates BST bounds
      return false; // @step:found
    }

    // Recurse: left subtree values must be less than current node
    // Right subtree values must be greater than current node
    return (
      validate(node.left, minVal, node.value) && // @step:search-node
      validate(node.right, node.value, maxVal) // @step:search-node
    );
  }

  return validate(root, -Infinity, Infinity); // @step:complete
}

File: src/algorithms/trees/bst-operations/bst-validation/sources/bst-validation.go
// BST Validation (Recursive) — validate BST property using min/max bounds

package main

import "math"

type BSTNode struct {
	value int
	left  *BSTNode
	right *BSTNode
}

func validate(node *BSTNode, minVal int, maxVal int) bool {
	if node == nil {
		return true // @step:initialize
	}

	if node.value <= minVal || node.value >= maxVal {
		// Node value violates BST bounds
		return false // @step:found
	}

	// Recurse: left subtree values must be less than current node
	// Right subtree values must be greater than current node
	return validate(node.left, minVal, node.value) && // @step:search-node
		validate(node.right, node.value, maxVal) // @step:search-node
}

func bstValidation(root *BSTNode) bool {
	return validate(root, math.MinInt64, math.MaxInt64) // @step:complete
}
